docs/architecture/adr-063-core-module-api.md

```diff
@@ -0,0 +1,546 @@
+stores - regular kv-store, memory, and transient:
+
+```go
+type KVStoreService interface {
+    OpenKVStore(context.Context) KVStore
+}
+
+type MemoryStoreService interface {
+    OpenMemoryStore(context.Context) KVStore
+}
+type TransientStoreService interface {
+    OpenTransientStore(context.Context) KVStore
+}
+```
+
+Modules can use these services like this:
+```go
+func (k msgServer) Send(ctx context.Context, msg *types.MsgSend) (*types.MsgSendResponse, error) {
+    store := k.kvStoreSvc.OpenKVStore(ctx)
+}
+```
+
+Just as with the current runtime module implementation, modules will not need to explicitly name these store keys,
+but rather the runtime module will choose an appropriate name for them and modules just need to request the
+type of store they need in their dependency injection (or manual) constructors.
+
+#### Event Service
+
+The event `Service` and `Manager` will be defined in the `cosmossdk.io/core/event` package.
+
+The event `Service` gives modules access to an event `Manager` which allows modules to emit typed and legacy,
+untyped events:
+
+```go
+package event
+type Service interface {
+    GetManager(context.Context) Manager
+}
+
+type Manager interface {
+    // Emit emits events to both clients and state machine listeners. These events MUST be emitted deterministically
+    // and should be assumed to be part of blockchain consensus.
+    Emit(proto.Message) error
+    
+    // EmitLegacy emits legacy untyped events to clients only. These events do not need to be emitted deterministically
+    // and are not part of blockchain consensus.
+    EmitLegacy(eventType string, attrs ...LegacyEventAttribute) error
+
+    // EmitClientOnly emits events only to clients. These events do not need to be emitted deterministically
+    // and are not part of blockchain consensus.
+    EmitClientOnly(proto.Message) error
+}
+```
+
+Typed events emitted with `Emit` can be observed by other modules and thus must be deterministic and should be assumed
+to be part of blockchain consensus (whether they are part of the block or app hash is left to the runtime to specify).
+
+Events emitted by `EmitLegacy` and `EmitClientOnly` are not considered to be part of consensus and cannot be observed
+by other modules. If there is a client-side need to add events in patch releases, these methods can be used.
+
+Design questions:
+* should we allow overriding event managers (like `sdk.Context.WithEventManager`)?
+
+#### Block Info Service
+
+The block info `Service` will be defined in the `cosmossdk.io/core/blockinfo` package.
+
+The `BlockInfo` service allows modules to get a limited set of information about the currently executing block
+that does not depend on any specific version of Tendermint:
+```go
+package blockinfo
+
+type Service interface {
+    GetBlockInfo(ctx context.Context) BlockInfo
+}
+
+type BlockInfo struct {
+    ChainID string
+    Height int64
+    Time time.Time
+    Hash []byte
```

Comment:
still don't know what hash this is? 

lastresulthash, app hash, last validator hash?